x/stake/keeper/slash.go

```diff
@@ -4,56 +4,176 @@ import (
 	"fmt"
 
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	types "github.com/cosmos/cosmos-sdk/x/stake/types"
 	crypto "github.com/tendermint/go-crypto"
 )
 
-// NOTE the current slash functionality doesn't take into consideration unbonding/rebonding records
-//      or the time of breach. This will be updated in slashing v2
-// slash a validator
-func (k Keeper) Slash(ctx sdk.Context, pubkey crypto.PubKey, height int64, fraction sdk.Rat) {
+// slash an unbonding delegation
+func (k Keeper) slashUnbondingDelegation(ctx sdk.Context, unbondingDelegation types.UnbondingDelegation, height int64, fraction sdk.Rat) (slashAmount sdk.Rat) {
+	now := ctx.BlockHeader().Time
+
+	// If unbonding started before this height, stake didn't contribute to infraction
+	if unbondingDelegation.CreationHeight < height {
+		return sdk.ZeroRat()
+	}
+
+	if unbondingDelegation.MinTime < now {
+		// Unbonding delegation no longer eligible for slashing, skip it
+		// TODO Settle and delete it automatically?
+		return sdk.ZeroRat()
+	}
+
+	// Calculate slash amount proportional to stake contributing to infraction
+	slashAmount = sdk.NewRatFromInt(unbondingDelegation.InitialBalance.Amount, sdk.OneInt()).Mul(fraction)
+
+	// Don't slash more tokens than held
+	slashAmountInt := slashAmount.EvaluateInt()
+	if slashAmountInt.GT(unbondingDelegation.Balance.Amount) {
+		slashAmountInt = unbondingDelegation.Balance.Amount
+	}
+
+	// Update unbonding delegation
+	unbondingDelegation.Balance.Amount = unbondingDelegation.Balance.Amount.Sub(slashAmountInt)
+	k.SetUnbondingDelegation(ctx, unbondingDelegation)
+
+	return
+}
+
+// slash a redelegation
+func (k Keeper) slashRedelegation(ctx sdk.Context, redelegation types.Redelegation, height int64, fraction sdk.Rat) (slashAmount sdk.Rat, tokensToBurn int64) {
+	now := ctx.BlockHeader().Time
+
+	// If redelegation started before this height, stake didn't contribute to infraction
+	if redelegation.CreationHeight < height {
+		return sdk.ZeroRat(), 0
+	}
+
+	if redelegation.MinTime < now {
+		// Redelegation no longer eligible for slashing, skip it
+		// TODO Delete it automatically?
+		return sdk.ZeroRat(), 0
+	}
+
+	// Calculate slash amount proportional to stake contributing to infraction
+	slashAmount = sdk.NewRatFromInt(redelegation.InitialBalance.Amount, sdk.OneInt()).Mul(fraction)
```

Comment:
redelegation doesn't have an initial balance only an initial shares... I guess we want to store the initial balance instead?

Reply:
aha I see you've added it in